File: bayesmendel-comparator/python/tests/test_presenter.py
import io
import json

from bayesmendel_comparator.comparison_builder import build_comparison
from bayesmendel_comparator.presenter import (
    present_csv,
    present_json,
    present_table,
    presenter_for,
)
from tests.fixtures_loader import load_all_fixtures


def _render(presenter: object, *, format_name: str = "") -> str:
    sink = io.StringIO()
    # Allow passing either a function directly or a format name via presenter_for.
    chosen = presenter if format_name == "" else presenter_for(format_name)  # type: ignore[arg-type]
    chosen(build_comparison(load_all_fixtures()), sink)  # type: ignore[operator]
    return sink.getvalue()


def test_table_writes_header_and_one_row_per_model() -> None:
    lines = _render(present_table).splitlines()

    assert len(lines) == 4
    assert "Model" in lines[0]
    assert "BRCAPRO" in lines[1]
    assert "MMRpro" in lines[2]
    assert "PancPRO" in lines[3]


def test_table_renders_missing_cells_as_dash() -> None:
    pancpro_line = _render(present_table).splitlines()[3]

    # PancPRO does not populate the BRCA1 column -- it must show as "-".
    assert " - " in pancpro_line


def test_csv_emits_header_and_one_row_per_model() -> None:
    lines = _render(present_csv).splitlines()

    assert len(lines) == 4
    assert lines[0].split(",")[0] == "Model"
    assert lines[1].split(",")[0] == "BRCAPRO"


def test_csv_does_not_quote_normal_strings() -> None:
    output = _render(present_csv)

    assert '"BRCAPRO"' not in output


def test_json_round_trips_to_columns_and_rows() -> None:
    parsed = json.loads(_render(present_json))

    assert "columns" in parsed
    assert "rows" in parsed
    assert len(parsed["rows"]) == 3
    assert parsed["rows"][0]["Model"] == "BRCAPRO"


def test_json_keeps_numeric_probabilities_as_numbers() -> None:
    parsed = json.loads(_render(present_json))

    assert parsed["rows"][0]["Pr(BRCA1 mutation)"] == 0.4239


def test_presenter_for_dispatches_by_name() -> None:
    assert presenter_for("table") is present_table
    assert presenter_for("csv") is present_csv
    assert presenter_for("json") is present_json

File: bayesmendel-comparator/r/tests/testthat/test-presenter.R
capture_output <- function(presenter, comparison) {
  sink <- textConnection("captured", "w", local = TRUE)
  on.exit(close(sink), add = TRUE)
  presenter(comparison, sink)
  paste(captured, collapse = "\n")
}

test_that("table presenter writes header and one row per model", {
  output <- capture_output(present_table, build_comparison(load_all_fixtures()))
  lines <- strsplit(output, "\n", fixed = TRUE)[[1L]]

  expect_equal(length(lines), 4L)
  expect_true(grepl("Model", lines[[1L]], fixed = TRUE))
  expect_true(grepl("BRCAPRO", lines[[2L]], fixed = TRUE))
  expect_true(grepl("MMRpro", lines[[3L]], fixed = TRUE))
  expect_true(grepl("PancPRO", lines[[4L]], fixed = TRUE))
})

test_that("table presenter renders missing cells as '-'", {
  output <- capture_output(present_table, build_comparison(load_all_fixtures()))
  pancpro_line <- strsplit(output, "\n", fixed = TRUE)[[1L]][[4L]]

  # PancPRO has no BRCA1 probability -- it must show as "-", not "0.00%".
  expect_true(grepl(" - ", pancpro_line, fixed = TRUE))
})

test_that("csv presenter emits header plus one row per model with commas", {
  output <- capture_output(present_csv, build_comparison(load_all_fixtures()))
  lines <- strsplit(output, "\n", fixed = TRUE)[[1L]]

  expect_equal(length(lines), 4L)
  expect_equal(strsplit(lines[[1L]], ",", fixed = TRUE)[[1L]][[1L]], "Model")
  expect_equal(strsplit(lines[[2L]], ",", fixed = TRUE)[[1L]][[1L]], "BRCAPRO")
})

test_that("csv presenter quotes fields containing commas", {
  output <- capture_output(present_csv, build_comparison(load_all_fixtures()))

  # The lifetime summary uses "; " between entries, but cells also include
  # percentages with commas if locale changed -- our format_cell always uses
  # a decimal point, so no comma should appear *inside* a cell, so no quoting
  # expected here.  Assert we haven't spuriously quoted normal strings:
  expect_false(grepl('"BRCAPRO"', output, fixed = TRUE))
})

test_that("json presenter emits columns and rows arrays", {
  output <- capture_output(present_json, build_comparison(load_all_fixtures()))
  parsed <- jsonlite::fromJSON(output, simplifyVector = FALSE)

  expect_true("columns" %in% names(parsed))
  expect_true("rows" %in% names(parsed))
  expect_length(parsed$rows, 3L)
  expect_equal(parsed$rows[[1L]]$Model, "BRCAPRO")
})

test_that("json presenter keeps numeric probabilities as numbers", {
  output <- capture_output(present_json, build_comparison(load_all_fixtures()))
  parsed <- jsonlite::fromJSON(output, simplifyVector = FALSE)

  expect_equal(parsed$rows[[1L]][["Pr(BRCA1 mutation)"]], 0.4239)
})

test_that("presenter_for returns the right function", {
  expect_identical(presenter_for("table"), present_table)
  expect_identical(presenter_for("csv"), present_csv)
  expect_identical(presenter_for("json"), present_json)
  expect_error(presenter_for("yaml"))
})
